docs/en/docs/settings.md
# Settings

Almost all settings are specified through environment variables, except server network related settings.

For example:
=== "Linux/MacOS"
    You can add it before the command
    ``` bash
    CONTENT=../mycomics/ thumb=/tmp/thumb/ comiclib
    ```
    Or set it first
    ``` bash
    export CONTENT=../mycomics/
    export thumb=/tmp/thumb/
    comiclib
    ```
=== "Windows"
    ```
    set CONTENT=..\mycomics\
    set thumb=C:\Users\Administrator\AppData\Local\Temp\thumb\
    comiclib
    ```
Environment variable keys are generally not case-sensitive.

The following is a list of available settings:

| Environment variable | Description | Default value |
| ------- | ---- | ----- |
| `debug` | Turn on debug output (`True`/`False`) | `False` |
| `loglevel` | Log level (`DEBUG`/`INFO`/`WARNING`/`ERROR`/`CRITICAL`). If `debug` is `True`, it will be overwritten to `DEBUG` | `INFO` |
| `content` | The path where the comic file is stored | `.` |
| `cover` | The path where the generated cover thumbnails are stored | `./thumb`|
| `thumb` | The path where the generated page thumbnails are stored. If not provided (`None`), will be the same as `cover`. | `None`|
| `thumb_width` | Maximum thumbnail width | 250 |
| `thumb_height` | Maximum thumbnail height | 350 |
| `metadata` | The URL for metadata database, refer to [SQLAlchemy documentation](https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls) | `sqlite:///./comiclib_metadata.db` |
| `password` | Admin password (also used as API Key currently) [^1]. If it is `None`, any visitor will have editing permissions. This feature is designed to protect against gentlemen but not villains. If you need security protection, please use e.g. the HTTP basic authentication of the reverse proxy, Cloudflare Access or TLS client certificate, etc. | `None`|
| `skip_exists`| Skip comics that have been scanned into the metadata database during scanning? (`True`/`False`) | `True` |
| `watch` | Monitor comic folders and automatically scan (`True`/`False`) | `True` |
| `display_title_format` | This string specifies the format of the display title.  `{title}`, `{subtitle}`, `{path}` can be used as a placeholder. `\n` will be escaped as a newline. Title cannot be edited while this is set. | `None` (original title) |
| `UA_convert_jxl` | For requests with matched user-agent, convert JPEG XL files to other popular formats on the server side. The value is a regular expression. | `Android` |
| `UA_convert_all` | For requests with matched user-agent, convert all files to other popular formats on the server side. The value is a regular expression. | `\b\B` (will not match anything) |

In addition, there are settings for scanners, please refer to the respective instructions.

[^1]: The security system may change in future versions, including the meaning of the environment variable `password`.


docs/zh/docs/settings.md
# 设置

几乎所有设置都通过环境变量指定，除了服务器网络相关设置。

例如：
=== "Linux/MacOS"
    你可以在命令前添加它
    ``` bash
    CONTENT=../mycomics/ thumb=/tmp/thumb/ comiclib
    ```
    或者先设定好
    ``` bash
    export CONTENT=../mycomics/
    export thumb=/tmp/thumb/
    comiclib
    ```
=== "Windows"
    ```
    set CONTENT=..\mycomics\
    set thumb=C:\Users\Administrator\AppData\Local\Temp\thumb\
    comiclib
    ```
环境变量的键一般不区分大小写。

下面是可用的设置列表：

| 环境变量 | 说明 | 默认值 |
| ------- | ---- | ----- |
| `debug` | 开启调试输出（`True`/`False`） | `False` |
| `loglevel` | 日志级别（`DEBUG`/`INFO`/`WARNING`/`ERROR`/`CRITICAL`），若`debug`为`True`，会被覆盖为`DEBUG` | `INFO` |
| `content` | 漫画文件存放的路径 | `.` |
| `cover` | 生成的封面缩略图存放的路径。 | `./thumb`|
| `thumb` | 生成的页面缩略图存放的路径。如果未提供 (`None`)，会和 `cover` 的值一样。| `None`|
| `thumb_width` | 缩略图的最大宽度 | 250 |
| `thumb_height` | 缩略图的最大高度 | 350 |
| `metadata` | 元数据库 URL，参考[SQLAlchemy 文档](https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls) | `sqlite:///./comiclib_metadata.db` |
| `password` | 管理密码（目前也用作 API 密钥）[^1]，若为`None`则任何访客皆可编辑。此功能防君子不防小人，若需安全保护请借助反向代理的 HTTP 基本验证、Cloudflare Access 或 TLS 客户端证书等。| `None`|
| `skip_exists`| 扫描时是否跳过曾扫入元数据库的漫画？（`True`/`False`）| `True` |
| `watch` | 监视漫画文件夹，自动扫描 （`True`/`False`）| `True` |
| `display_title_format` | 该字符串指定展示标题的格式。可以使用 `{title}`, `{subtitle}`, `{path}` 作为占位符，`\n` 会被转义为换行。设定此项时无法编辑标题。 | `None`（原标题） |
| `UA_convert_jxl` | 对于哪些 user-agent 的请求在服务端将 JPEG XL 文件转为其他流行格式，该值是一个用于匹配的正则表达式 | `Android` |
| `UA_convert_all` | 对于哪些 user-agent 的请求在服务端将所有文件转为其他流行格式，该值是一个用于匹配的正则表达式 | `\b\B`（不匹配任何东西）|

此外还有扫描脚本的设置，请参看各自的说明。

[^1]: 安全系统可能在未来的版本变更，包括环境变量 `password` 的含义。


comiclib/config.py
from pydantic_settings import BaseSettings
from typing import Union, Optional
import re
import logging

logger = logging.getLogger('ComicLib')

class Settings(BaseSettings):
    debug: bool = False
    loglevel: str = 'INFO'
    content: str = '.'
    cover: str = './thumb'
    thumb: Union[str, None] = None
    thumb_width: int = 250
    thumb_height: int = 350
    metadata: str = 'sqlite:///./comiclib_metadata.db'
    password: Union[str, None] = None
    skip_exits: bool = True
    watch: bool = True
    display_title_format: Optional[str] = None
    display_subtitle: Optional[bool] = None
    UA_convert_jxl: str = 'Android'
    UA_convert_all: str = r'\b\B'  # default: match nothing
    skip_scan_error: bool = False

settings = Settings()

if settings.debug:
    settings.loglevel = 'DEBUG'
numeric_level = getattr(logging, settings.loglevel.upper(), None)
if not isinstance(numeric_level, int):
    raise ValueError(f'Invalid log level: {settings.loglevel}')
logging.basicConfig(format='%(asctime)s %(levelname)-8s %(name)s: %(message)s',
                    datefmt='%Y-%m-%d %H:%M:%S',
                    level=numeric_level)
logger.debug(settings)

if settings.display_subtitle is not None:
    logger.warning('The environment variable display_subtitle is deprecated and will be removed in the future. Please use display_title_format instead.')

settings.UA_convert_jxl = re.compile(settings.UA_convert_jxl)
settings.UA_convert_all = re.compile(settings.UA_convert_all)
if settings.thumb is None:
    settings.thumb = settings.cover
